frontend/src/javascript/utils/codeUtils.js

```diff
@@ -0,0 +1,95 @@
+import store from "@/store";
+
+export function compareCode(oldText, newText, event, selection) {
+    let oldLen = oldText.length;
+    let newLen = newText.length;
+    let lengthDifference = newLen - oldLen;
+
+    let oldLineCount = oldText.numberOfLines();
+    let newLineCount = newText.numberOfLines();
+    let lineCountDifference = newLineCount - oldLineCount;
+
+    let ret = {
+        start: selection.start,
+        end: selection.end,
+        size: lengthDifference,
+        deltaLineCount: lineCountDifference,
+    };
+
+    if (selection.start != selection.end) return ret;
+    if (lengthDifference >= 0) return ret;
+
+    if (!event.inputType.includes("Backward") && !event.inputType.includes("Forward")) {
+        console.error("Unhandled input type: ", event.inputType);
+        return ret;
+    }
+
+    if (event.inputType.includes("Backward")) {
+        ret.start -= Math.abs(lengthDifference);
+    } else {
+        ret.end += Math.abs(lengthDifference);
+    }
+
+    return ret;
+}
+
+export function moveTrackedVariables(changes) {
+    const handleVarTrackerMove = (ch, varObj) => {
+        if (varObj == null) return;
+
+        if (ch.end > varObj.start && ch.start < varObj.end) {
+            if (ch.start >= varObj.start && ch.end <= varObj.end) {
+                varObj.end += ch.size;
+                if (varObj.end - varObj.start <= 0) return "Delete";
+                return;
+            }
+            return "Delete";
+        }
+
+        if (ch.end <= varObj.start) {
+            varObj.start += ch.size;
+            varObj.end += ch.size;
+        }
+    };
+
+    store.dispatch("project/filterSceneObjects", (sceneObj) => {
+        let mainVar = sceneObj.variable;
+        if (handleVarTrackerMove(changes, mainVar) == "Delete") {
+            return false;
```

Comment:
Tą główną zmienną też bym zrobił na tej samej zasadzie co podobiekty. czyli, zeby ją usuwać z obiektu, a le sam obiekt zostawić. Na scenie w górnym lewym rogu trzeba by zmienić, że jak brakuje tej zmiennej to w miejsce jej nazwy wyświetlamy, np. napis "null". Tak wgl ta funkcja wykonuje się dla metody filter i to nie jest dobrą praktyką, żeby w filtrach modyfikować obiekty.